### PR TITLE
`Programming exercises`: Add Gradle as project type for Java exercises on Atlassian setup

### PR DESCRIPTION
**Pull request by ole-ve**
### Checklist
This PR has quite a lot of line additions but this is mainly due to the newly added template files (like .gitignore, wrapper files, etc.). There are around 250 added lines of Java-/TypeScript-Code.
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple screenshots/screencasts of my UI changes.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).
-> The new project type is only available on the Atlassian Setup. However, these changes should not cause any problems on the Jenkins/Gitlab-Setup. In addition, these project types should not be selectable on this setup.

### Motivation and Context
As part of the Bachelor Thesis from @Hialus and @ole-ve about the automatic generation of code hints for programming exercises (HESTIA) in Java, testwise coverage has to be recorded. We will use a plugin that is only available for Gradle projects, therefore we decided to add Gradle as a new project type.
In addition, Maven is getting less innovative and may be considered as not state-of-the-art anymore.

### Description
This PR introduces two new project types:
- Plain - Gradle: The template and solution repositories are plain Java code, the test repository is a Gradle project. The Maven equivalent in Artemis is 'Eclipse' for which the template and solution repositories are also plain Java code but the test repository is a Maven project.
- Gradle - Gradle: The template, solution, and test repository are all Gradle projects. The Maven equivalent is 'Maven' where all three repositories are Maven projects

The creation of the project files has been adapted as well as the creation of the build plans on the Bamboo server.

### Related changes
In order to work, the following related changes have been made:
- Changing the Ares testing framework to support Gradle: [#191](https://github.com/ls1intum/Ares/pull/191)
- Cache the Gradle dependency in the artemis-maven-docker which is used for executing the builds on Bamboo/Jenkins: [#63](https://github.com/ls1intum/artemis-maven-docker/pull/63)

### Steps for Testing
Prerequisites:
- Testserver with the Atlassian Setup (TS1, TS3)
- 1 Instructor
- 1 Student

1. Log in to Artemis
2. Navigate to Course Administration
3. Navigate to Exercises
4. Generate new Java programming exercises. Check for both projects type 'Plain - Gradle' and 'Gradle - Gradle' that the following things are working correctly:
- creating the build plans and repositories
- Template Score: 0%, Solution Score: 100%
- importing as a Gradle project into the local IDE works fine
- when activating the static code analysis, introducing some style issues, and pushing the changes, the build result contains feedback on these style issues
- when applying and pushing changes to (one of the three) repositories, the test case results update accordingly and these results are as expected
- when pushing some code that does not compile, the build result shows a build error
- adding new dependencies to the build.gradle in the test repository and using classes from this dependency works fine
- when activating sequential test runs, confirm that the behavioral tests are only executed after the structural tests have all passed. Otherwise, no test results are shown for these behavioral tests.
- clone all three repositories locally. Import the test repository. Add the template and solution repositories as an existing module. Navigate in IntelliJ to "File" -> "Project Structure" -> "Modules" -> "$ProjectName-Tests" -> Sub-Option "test" -> "Dependencies" -> Button "+" -> "Module Dependency" -> Select the template or solution repository with appended ".main". Verify now that the Test classes "recognize" the code from the respective (template or solution) repository. Execute "./gradlew clean test" and check if all tests run as expected.
 
Only for project type 'Gradle - Gradle':
- Check that also the solution and template repository are Gradle projects
- Add a dependency to all three repositories and use classes from this dependency. When pushing the changes the result should be as expected.

5. Code editor: Build errors and style issues are displayed with a symbol in the corresponding line in the respective file.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots
Bildschirmfoto 2022-03-10 um 20 04 39
Bildschirmfoto 2022-03-10 um 20 04 49

| File | Branch | Line |
|------------|-------:|-----:|
| ProgrammingExerciseService.java | 81% | 90.04% |
| FeedbackRepository.java | 81% | 86.9% |
| FileService.java | 86% | 89.7% |

### CODE REVIEW

File: src/main/resources/templates/java/gradle_gradle/exercise/git.ignore.file
Comment (Hialus) [inline src/main/resources/templates/java/gradle_gradle/exercise/git.ignore.file:92]: Is there any specific reason why you exclude all Gradle wrapper files except `gradle/wrapper/gradle-wrapper.jar` here? (And in the other git ignore files as well)
Usually, the Gradle wrapper is shipped together with the repository. It would also make sense IMO to put the gradle wrapper in the template and solution repos, as this would make the Execution of the code easier.

Unless Stephan has anything against it I think we should include the gradle wrapper in the template.

File: src/main/resources/templates/java/test/gradle/projectTemplate/readme.md
Comment (Hialus) [inline src/main/resources/templates/java/test/gradle/projectTemplate/readme.md:4]: Yeah. This won't work unless you include the Gradle wrapper.
If you do not include it, it has to be created first. (Which you should explain here)

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java
Comment (Hialus) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java:506]: It might make sense to extract this and the related check for Maven to helper methods (Or maybe methods in the enum?), as these conditions are checked multiple times in this service and the BambooBuildPlanService.
Not strictly necessary, but it would increase the readability of the code IMO.

**Review by Hialus — CHANGES_REQUESTED** [review]
The code looks good overall. I just have one suggestion and one issue with the template.
I would also like to see some server tests for the new changes.

File: src/main/resources/templates/java/gradle_gradle/exercise/git.ignore.file
Comment (ole-ve) [reply src/main/resources/templates/java/gradle_gradle/exercise/git.ignore.file:92]: This is a relict and I will change to exclude all wrapper files.
We will not use the wrapper but the `gradle` command instead directly. This has to do with the way the builds are executed on Bamboo in the corresponding Docker container.
In case the wrapper is required, e.g. to run tests with a specific Java/Gradle version, the .gitignore can still be edited and the build plan can be changed accordingly.

File: src/main/resources/templates/java/test/gradle/projectTemplate/readme.md
Comment (ole-ve) [reply src/main/resources/templates/java/test/gradle/projectTemplate/readme.md:4]: Changed to `gradle clean test`.

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java
Comment (ole-ve) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java]: I added methods to the enum.

**Comment by ole-ve** [comment]
> I would also like to see some server tests for the new changes.

I am working on that. I am figuring out a neat way how to execute Gradle builds from within Java.

**Comment by dfuchss** [comment]
This seems to be interesting!
Do you also plan to create build plans for Jenkins (I would assume that this should be easily doable :) and would enhance the PR a lot) ?

**Comment by ole-ve** [comment]
> This seems to be interesting! Do you also plan to create build plans for Jenkins (I would assume that this should be easily doable :) and would enhance the PR a lot) ?

This should be no big challenge and I will definitely implement it. But maybe not in this PR but as a soon (and very small) follow-up since this PR already adds a lot of functionality to be tested into a sensible area of Artemis.

**Comment by b-fein** [comment]
I experimented a bit with the new project type:
- The problem statement contains for example the test case `testBubbleSort`. This always shows up in grey in the problem statement as the result is reported for test case `testBubbleSort()`. The same applies to other behavioural test cases.
- Code issues are not marked in the online editor as the path contains the prefix `/opt/bambooagent/bamboo-agent-home/xml-data/build-dir/TESTINGSESSIONGRADLEBFEIN-SOLUTION-JOB1/` instead of e.g., only `src/main/resources/bfein/Context.java`.
- The build errors are not marked in the code editor but only shown in the build output. The current parsing code expects the Maven format for errors.
- Locally, running `./gradlew` does not seem to work because the actual `gradle/wrapper/gradle-wrapper.jar` is missing.
- I think it would be nice to have another entry in the `sourceSets` of `build.gradle`
  ```groovy
  test {
      java {
          srcDir "localtest" // not necessarily final name
      }
  }

  ```
  so that students can easily add their own unit tests during local development if they want. That `localtest` directory would have to be created beforehand and contain some `.keep`/`.gitkeep` file so that the Alt+Insert->Test action works correctly in IntelliJ.

**Comment by b-fein** [comment]
Okay, I only suggested it because we have got some exercises where we encourage writing own tests.
I looks like the default Maven template does not have a local tests directory set up either, so I assume it is not/rarely used at TUM. Therefore, not adding it to this template either and requiring the additional small setup step for instructors is perfectly fine by me.

**Comment by ole-ve** [comment]
> * The problem statement contains for example the test case `testBubbleSort`. This always shows up in grey in the problem statement as the result is reported for test case `testBubbleSort()`. The same applies to other behavioural test cases.

79b4abf: I changed the problem statement for the gradle template accordingly.

> * Code issues are not marked in the online editor as the path contains the prefix `/opt/bambooagent/bamboo-agent-home/xml-data/build-dir/TESTINGSESSIONGRADLEBFEIN-SOLUTION-JOB1/` instead of e.g., only `src/main/resources/bfein/Context.java`.

2657b1f: Directories are included in the static code analysis that are not part of the copied submission (`src/main/java` and `src/main/resource`). I implicitly removed those directories from the main source set by setting the assignment folder (where the copied submission is located) as the only main source. There is no scenario where files in these directories would be used in the test repository.

> * The build errors are not marked in the code editor but only shown in the build output. The current parsing code expects the Maven format for errors.

487f920: I fixed this by introducing a new regex matching the build error for Gradle build output. I know that the taken solution is not the best, but this is mainly due to the messy code parsing the build log for errors. Refactoring is definitely required, but also out of scope for this PR. I will take a look at this in a separate PR.

> * Locally, running `./gradlew` does not seem to work because the actual `gradle/wrapper/gradle-wrapper.jar` is missing.

b30be0f: I fixed this by including the corresponding gradle-wrapper.jar into the repository.

File: src/main/resources/templates/java/gradle_gradle/readme
Comment (Hialus) [inline src/main/resources/templates/java/gradle_gradle/readme:13]: Optional change:
Instead of declaring them here, you could also filter out the empty brackets on the server-side. This would keep the test names consistent with Maven and make porting existing Maven to Gradle exercises easier.
But it is also fine to do it how you do it now 🙂

**Review by Hialus — CHANGES_REQUESTED** [review]
Reviewed the code.
Your changes look good overall. I just found some minor improvements.

Great job, especially considering that the code you are working with was not of the best quality, to begin with.

File: build.gradle
Comment (ole-ve) [reply build.gradle:141]: No, it is not. I will do that when he is available again. The actual reason, why the coverage decreases is that the template classes (BubbleSort.java, etc.) are also taken into these calculations which obviously makes no sense.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooBuildPlanService.java
Comment (ole-ve) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooBuildPlanService.java:182]: I would have to take a look into the live database just to be sure. But as far as I know, all Java exercises that have no project type (null) should be of project type Maven. I will double-check this and create an issue if possible.

File: build.gradle
Comment (Hialus) [reply build.gradle:141]: Could we maybe specify that Jacoco should ignore the path of the templates? 
Would make a lot more sense imo.

File: src/main/resources/templates/java/exercise/git.ignore.file
Comment (ole-ve) [reply src/main/resources/templates/java/exercise/git.ignore.file]: Yes, you are right. It actually does not make a difference because this file is overwritten in case Gradle is chosen as the project type for the template/solution repositories. But for the sake of consistency, I removed it.

File: src/main/resources/templates/java/gradle_gradle/readme
Comment (ole-ve) [reply src/main/resources/templates/java/gradle_gradle/readme:13]: We could do that. But in my opinion, this would cause the code to be more complex for this one project type. And given the fact, that porting exercises is a rare use case and that changing the test names manually can easily be done, I would refrain from doing this.

File: build.gradle
Comment (ole-ve) [reply build.gradle:141]: Yes, probably. I will discuss this with Stephan.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooBuildPlanService.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooBuildPlanService.java:303]:
```suggestion
    private Stage getRegularTestTaskForJavaAndKotlinExercise(Stage defaultStage, Job defaultJob, VcsCheckoutTask checkoutTask, boolean isMavenProject) {
```
I think they are usually called regular test runs instead of non-sequential.

File: src/main/resources/templates/java/gradle_gradle/exercise/gradle/wrapper/gradle-wrapper.properties
Comment (b-fein) [reply src/main/resources/templates/java/gradle_gradle/exercise/gradle/wrapper/gradle-wrapper.properties]: Maybe putting a readme document in `src/main/resources/templates/` next to the `error.html` would be fine without interfering with the actual templates.
Then the document could also be extended in the future by others with information about the other templates as needed. If it grows bigger, then it can always be moved to a proper documentation page in the `docs/dev/` directory (or you start a new page there directly).

File: src/main/resources/templates/java/test/gradle/projectTemplate/build.gradle
Comment (Strohgelaender) [inline src/main/resources/templates/java/test/gradle/projectTemplate/build.gradle:16]: Since Ares is only used in the tests, I think `testImplementation` should be enough here. Related https://github.com/ls1intum/Ares/issues/197
```suggestion
    testImplementation 'de.tum.in.ase:artemis-java-test-sandbox:1.9.0'
```

File: src/main/webapp/app/entities/build-log.model.ts
Comment (Strohgelaender) [inline src/main/webapp/app/entities/build-log.model.ts:57]:
```suggestion
    extractErrors(programmingLanguage?: ProgrammingLanguage, projectType?: ProjectType): Array<Annotation> {
```

File: src/main/webapp/app/shared/orion/orion-build-and-test.service.ts
Comment (Strohgelaender) [inline src/main/webapp/app/shared/orion/orion-build-and-test.service.ts:91]:
```suggestion
    private forwardBuildLogs(participationId: number, programmingLanguage?: ProgrammingLanguage, projectType?: ProjectType) {
```

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java
Comment (ole-ve) [reply src/main/java/de/tum/in/www1/artemis/service/FileService.java:581]: I agree that this would be better style. But because of the recursive way the method is implemented, a lot of copy-pasting would be required that would outweigh the benefit of the overloading. What are your thoughts on that?

File: src/main/resources/templates/java/gradle_gradle/exercise/gradle/wrapper/gradle-wrapper.properties
Comment (ole-ve) [reply src/main/resources/templates/java/gradle_gradle/exercise/gradle/wrapper/gradle-wrapper.properties]: I created the README file in 230556a.

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java
Comment (b-fein) [reply src/main/java/de/tum/in/www1/artemis/service/FileService.java:581]: Couldn’t the overloaded method just be defined as
```java
public void replaceVariablesInFileRecursive(String startPath, Map<String, String> replacements) throws IOException {
    replaceVariablesInFileRecursive(startPath, replacements, List.of());
}
```
Then no copy-pasting is required.

File: src/main/resources/templates/README.md
Comment (b-fein) [inline src/main/resources/templates/README.md:8]:
```suggestion
some parts may be missing and will be extended in the future.
```

File: src/main/resources/templates/README.md
Comment (b-fein) [inline src/main/resources/templates/README.md:11]:
```suggestion
### Updating to the Newest Ares Version
```

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java
Comment (ole-ve) [reply src/main/java/de/tum/in/www1/artemis/service/FileService.java:594]: Good idea, implemented. 👍

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java:516]: `repository.getLocalPath().toAbsolutePath().toString()` looks wierd for me, is this really necceary / no better way?

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java:630]:
```suggestion
                        Files.copy(stagePomXml.getInputStream(), buildStagePath.resolve("pom.xml"));
```

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java
Comment (ole-ve) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java:516]: Why does this look weird for you? `getLocalPath` returns the path of the local repository location on the server instance, `toAbsolutePath` is required to be able to access the files independent of the current working directory.

File: src/main/java/de/tum/in/www1/artemis/repository/FeedbackRepository.java
Comment (ole-ve) [reply src/main/java/de/tum/in/www1/artemis/repository/FeedbackRepository.java:190]: It is necessary, because ProgrammingLanguage has no attribute ProjectType.

File: src/main/resources/templates/java/gradle_gradle/exercise/src/${packageNameFolder}/Client.java
Comment (Hialus) [inline src/main/resources/templates/java/gradle_gradle/exercise/src/${packageNameFolder}/Client.java:59]:
```suggestion
     * @throws ParseException if date string cannot be parsed
```

Comment (Hialus) [reply src/main/resources/templates/java/gradle_gradle/exercise/src/${packageNameFolder}/Client.java:59]: Also an issue in the other template files

File: src/main/resources/templates/java/plain_gradle/solution/src/${packageNameFolder}/Client.java
Comment (Hialus) [inline src/main/resources/templates/java/plain_gradle/solution/src/${packageNameFolder}/Client.java:53]: And here

**Review by mallgaier — DISMISSED** [review]
Tested with Plain Java + Gradle without SCA.
Works fine 👍

**Review by Hialus — APPROVED** [review]
Tested in testing session on TS1.
Gradle-Gradle with SCA works as expected.
You just overlooked the ParserException in two places

**Review by valeriebucher — DISMISSED** [review]

**Review by akesfeden — DISMISSED** [review]
Tested in Testing Session on TS1, no issues encountered.

**Comment by dfuchss** [comment]
Could you please link a follow-up issue / pr here that deals with gradle for jenkins/gitlab as well :)
At my university, some people are looking forward to this. Therefore, I could link such an issue to them :)

**Review by mallgaier — APPROVED** [review]

**Review by Hialus — APPROVED** [review]
Re-approve. (Test & Code) 👍🏻

**Review by akesfeden — APPROVED** [review]
Re-approve after merge.

**Review by valeriebucher — APPROVED** [review]
Re-approve👍

**Review by b-fein — APPROVED** [review]
Code changes since last review look good.
Also worked as intended in the testing session yesterday.

Re-approve.

**Pull request merged**
